### PR TITLE
Beta requirements for JobTrackingWithFinalizers

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind feature
/sig apps

#### What this PR does / why we need it:

Satisfies beta requirements for JobTrackingWithFinalizers (but it doesn't graduate the feature gate yet):

- Limit the number of pods tracked in a single sync to avoid starvation of small jobs.
- Add metric `job_pod_finished_total`, which increments when the job controller finished tracking a pod (finalizer removed and pod accounted in status counters)

#### Which issue(s) this PR fixes:

Ref kubernetes/enhancements#2307

#### Special notes for your reviewer:

Review only the last 2 commits.
Builds on top of #104666.

#### Does this PR introduce a user-facing change?

dropped release note with revert in https://github.com/kubernetes/kubernetes/pull/105181

```release-note
NONE
```

original release note content:
```
When feature gate JobTrackingWithFinalizers is enabled:
- Limit the number of pods tracked in a single job sync to avoid starvation of small jobs.
- The metric job_pod_finished_total counts the number of finished pods tracked by the job controller
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-tracking-with-finalizers
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/assign @logicalhan 
for metrics

/assign @soltysh

File: pkg/controller/job/metrics/metrics.go [hunk 1]
```diff
@@ -68,10 +68,26 @@ var (
...
+	// finished tracking.
+	// It only applies to Jobs that were created while the feature gate
+	// JobTrackingWithFinalizers was enabled.
+	// Possible label values:
+	//   completion_mode: Indexed, NonIndexed
+	//   result:          failed, succeeded
+	JobPodFinished = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem: JobControllerSubsystem,
+			Name:      "job_pod_finished_total",
```
Comment (jan--f) [inline pkg/controller/job/metrics/metrics.go:82]: Nit:
```suggestion
			Name:      "job_pods_finished_total",
```
Adding the `jobs` plural here seems more appropriate and consistent with the help text. I'd also :+1: changing the variable name accordingly.

**Review by jan--f — APPROVED** [review]
I have one minor naming suggestion, but overall this lgtm from the metrics perspective.

@ehashman would appreciate your eyes and official stamp of approval.

File: pkg/controller/job/metrics/metrics.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/job/metrics/metrics.go]: I wasn't sure about the most appropriate name. Thanks!

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -1547,3 +1551,13 @@ func ensureJobConditionStatus(list []batch.JobCondition, cType batch.JobConditio
...
 }
+
+func recordJobPodFinished(job *batch.Job, oldCounters batch.JobStatus) {
+	completionMode := completionModeStr(job)
+	if diff := job.Status.Succeeded - oldCounters.Succeeded; diff > 0 {
+		metrics.JobPodsFinished.WithLabelValues(completionMode, metrics.Succeeded).Add(float64(diff))
+	}
+	if diff := job.Status.Failed - oldCounters.Failed; diff > 0 {
+		metrics.JobPodsFinished.WithLabelValues(completionMode, metrics.Failed).Add(float64(diff))
+	}
```
Comment (dgrisonnet) [inline pkg/controller/job/job_controller.go:1562]: nit: we could directly `Add` the difference instead of checking that it is superior to zero.

**Review by dgrisonnet — COMMENTED** [review]
One nit from my side, otherwise it looks good from an instrumentation perspective.

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:1562]: done

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**